### ISSUE TITLE
Batch Normalization axis

### ISSUE THREAD

**Issue by jbaron** [pytorch/pytorch#21856]
## 🚀 Feature
Make it possible to specify the axis on which to perform Batch Normalization.

## Motivation
Right now batch normalization is geared towards convolution layers which have typically a  layout of NxCxWxH (in the context of 2D batch normalization). But there are other scenarios possible (for example to apply BN to a RNN) where the features are located at a different axis. The current solution doesn't cater for those scenarios.

## Pitch
Adding the axis makes batch normalization more generic applicable to a wide variety of use cases and is inline with other ML frameworks. Default could still be axis=1 so NxCxWxH is handled as is currently the case but other cases are now also supported.

## Alternatives
The only workaround right now it to shift the different axis before the BN and then move them back again after. Not the most elegant solution.

**Comment by SsnL** [comment]
In RNNs, you often want to use LayerNorm instead.

**Comment by umanwizard** [comment]
@cpuhrsch  any thoughts ?

**Comment by jbaron** [comment]
@SsnL: the case I had at hand required some BN before going into a LSTM layer (so not a "LayerNormLSTMCell" type of solution). But mainly I was a bit surprised that it could not be parametrized since I would expect also other use cases (but perhaps I'm wrong there).

**Comment by ezyang** [comment]
This seems like a reasonable ask, we'll accept a PR that implements this. @ifedan mentions that we had something similar going on with the `dist` operator as well.

**Comment by VitalyFedyunin** [comment]
Please contact me before writing code, me and @cpuhrsch had several thoughts how to use TensorIterator in batch/layer norm.
